common/scala/src/main/scala/whisk/core/database/cosmosdb/CosmosDBArtifactStoreProvider.scala

```diff
@@ -0,0 +1,104 @@
+import akka.stream.ActorMaterializer
+import com.microsoft.azure.cosmosdb.rx.AsyncDocumentClient
+import spray.json.RootJsonFormat
+import whisk.common.Logging
+import whisk.core.database._
+import pureconfig._
+import whisk.core.ConfigKeys
+import whisk.core.database.cosmosdb.CosmosDBUtil.createClient
+import whisk.core.entity.{DocumentReader, WhiskActivation, WhiskAuth, WhiskEntity}
+
+import scala.reflect.ClassTag
+
+case class CosmosDBConfig(endpoint: String, key: String, db: String)
+
+case class ClientHolder(client: AsyncDocumentClient) extends Closeable {
+  override def close(): Unit = client.close()
+}
+
+object CosmosDBArtifactStoreProvider extends ArtifactStoreProvider {
+  type DocumentClientRef = ReferenceCounted[ClientHolder]#CountedReference
+  private lazy val config = loadConfigOrThrow[CosmosDBConfig](ConfigKeys.cosmosdb)
+  private var clientRef: ReferenceCounted[ClientHolder] = _
+
+  override def makeStore[D <: DocumentSerializer: ClassTag](useBatching: Boolean)(
+    implicit jsonFormat: RootJsonFormat[D],
+    docReader: DocumentReader,
+    actorSystem: ActorSystem,
+    logging: Logging,
+    materializer: ActorMaterializer): ArtifactStore[D] = {
+    makeStoreForClient(config, useBatching, getOrCreateReference(config))
+  }
+
+  def makeStore[D <: DocumentSerializer: ClassTag](config: CosmosDBConfig, useBatching: Boolean)(
+    implicit jsonFormat: RootJsonFormat[D],
+    docReader: DocumentReader,
+    actorSystem: ActorSystem,
+    logging: Logging,
+    materializer: ActorMaterializer): ArtifactStore[D] = {
+
+    makeStoreForClient(config, useBatching, createReference(config).reference())
+  }
+
+  private def makeStoreForClient[D <: DocumentSerializer: ClassTag](config: CosmosDBConfig,
+                                                                    useBatching: Boolean,
+                                                                    clientRef: DocumentClientRef)(
+    implicit jsonFormat: RootJsonFormat[D],
+    docReader: DocumentReader,
+    actorSystem: ActorSystem,
+    logging: Logging,
+    materializer: ActorMaterializer): ArtifactStore[D] = {
+
+    val classTag = implicitly[ClassTag[D]]
+    val (dbName, handler, viewMapper) = handlerAndMapper(classTag)
+
+    new CosmosDBArtifactStore(dbName, config, clientRef, handler, viewMapper)
+  }
+
+  private def handlerAndMapper[D](entityType: ClassTag[D])(
+    implicit actorSystem: ActorSystem,
+    logging: Logging,
+    materializer: ActorMaterializer): (String, DocumentHandler, CosmosDBViewMapper) = {
+    entityType.runtimeClass match {
+      case x if x == classOf[WhiskEntity] =>
+        ("whisks", WhisksHandler, WhisksViewMapper)
+      case x if x == classOf[WhiskActivation] =>
+        ("activations", ActivationHandler, ActivationViewMapper)
+      case x if x == classOf[WhiskAuth] =>
+        ("subjects", SubjectHandler, SubjectViewMapper)
+    }
+  }
+
+  private def getOrCreateReference(config: CosmosDBConfig) = synchronized {
```

Comment:
I get the feeling this reference is special - can you outline the goals of `ReferenceCounted` and `synchronized`?

nm: I see the "Shutdown Handling" comment

Reply:
`ArtifactStore` has a `shutdown()` function - is there a reason we cannot use that instead of the specialized ref counting? (client is responsible for shutdown). I don't see any non-test usages of `shutdown` currently - what is the hazard, if any, in either treating this the same as couchdb (no shutdown called, except for test cases) OR requiring that clients invoke shutdown (e.g. via `sys.addShutdownHook(store.shutdown())` in controller/invoker)? 

Part of the diff in client handling from couchdb is the client sharing - you mention that stores _share the same AsyncDocumentClient instance_, but it's not clear why?

Just looking for ways to simply this

Reply:
After reading some of the cosmosdb doco (sorry, cosmos n00b here), I see now that the "databases" are modeled as "multiple collections within a single db" which partly explains the client sharing, but I'm still wondering what is the impact if we create per-collection clients?

Reply:
Each `AsyncDocumentClient` encapsulates a HttpClient which own a pool of HTTP Connections to CosmosDB. As for all 3 collections we are connecting to same url at backend it would be better to share the pool.

Compared to this in `CouchDBRestStore` the pool is managed on per `ActorSystem` level which by design is then shared across various `ArtifactStore` implementations. In [PoolingRestClient#shutdown][1] method you would see a note around not closing the pool assuming shutdown is only relevant for tests. So ref counted approach used in this PR is meant to solve that 

[1]: https://github.com/apache/incubator-openwhisk/blob/master/common/scala/src/main/scala/whisk/http/PoolingRestClient.scala#L150

Reply:
getOrCreateReference is called once at create time - is there concurrency here in making the stores that requires synchronization or is this mostly defensive to get the counts right?

Reply:
can you add a comment explaining the synchronized fwiw.